up-l1/README.adoc

```diff
@@ -50,157 +42,218 @@ We will use the following terminology during this document:
+`T` is the type of the operation's successful outcome. `E` is the type of the operation's erroneous outcome.
 
 |===
 
- * <<UStatus>> is defined in link:../basics/error_model.adoc[uProtocol Error Model]
-
-
-NOTE: uProtocol data model used by the uTransport interface are declared in https://github.com/eclipse-uprotocol/up-core-api[up-core-api]
+NOTE: The data types used in the following sections are defined in link:../basics/README.adoc[uProtocol Basic Types].
 
 == UListener
 
-`UListener` is an interface used for receiving messages by the client uE from the uTransport layer. Listeners are used for receiving messages asynchronously by the uTransport layer.  
+A uEntity registers a `UListener` with the transport in order to process (incoming) messages that are of interest to the uEntity.
 
-  * Type name *MUST* be `UListener`
-  * *MUST* contain the method `void onReceive(UMessage&)` that is called by the Transport to notify (callback) the client
+=== OnReceive
 
+A transport invokes this method for each newly arrived message that matches the criteria specified as part of <<_registerlistener,registering the listener>>.
 
-== UTransport
+[source]
+----
+onReceive(IN message: UMessage&)
+----
 
-The transport layer API responsible for point-2-point communication. The interface provides common functionality across all transport implementations.
+==== Parameters
+.OnReceive Parameters
+[width="100%",cols="15%,15%,70%"]
+|===
+|Parameter | Type | Description
 
-* *MUST* implement all APIs defined in this section
+| message
+| link:../basics/umessage.adoc[UMessage]
+| The newly received message.
 
-=== Send()
+|===
 
-Transmit a UMessage.  
 
+== UTransport
 
-Communication protocols (a.k.a. uProtocol Transports) define their own Protocol Data Unit (PDU) that comprises of header and payload. Some transports header parameters map already to uProtocol UAttributes. Sending UAttributes in the transport's payload as well as in the header increases overhead for little to no gain. To address this issue,  <<Send()>> allows flexibility for uTransport implementers to define however they see fit what gets mapped into their transport header vs payload. 
+This is the main entry point into a transport's messaging functionality.
 
-For example, if HartleyTransport can map only UPriority to its PDU header, then HartleyTransport would define a PDU payload type that includes UPayload and all the other UAttributes and send that new type in its PDU payload such that the receiver will not loose any metadata. 
+=== Send
 
-The API signature:
+Clients use this method to transmit a single message.
 
-`OUT UStatus send(UMessage&)`
+[source]
+----
+send(
+  IN message: UMessage&,
+  OUT result: Future<Void, UStatus>
```

Comment:
It is in the Rust library :-)
I wonder if we should actually bother with sync vs async in this spec at all or not leave this up to the language library developers instead.

Reply:
Transport I think we need to be explicit as we said historically that send() returning that the message has been delivered to the "next hop". It is only the transport layer acknowledgement and not the communication layer (ex. rpc or pub/sub) acknowledgement.

Reply:
> Transport I think we need to be explicit as we said historically that send() returning that the message has been delivered to the "next hop". It is only the transport layer acknowledgement and not the communication layer (ex. rpc or pub/sub) acknowledgement.

And I think that this is the important thing to make clear. Whether an implementation implements the delivery to the _next hop_ using a blocking/synchronous call or a non-blocking/asynchronous call does not really make a difference regarding the delivery semantics/guarantees, does it?

Reply:
Whether it uses blocking or not underneath uTransport, the API itself has to be consistent, otherwise we cannot build a generic uStreamer that behaves the same regardless of the underlining transport implementation.

Reply:
Agreed, we do need consistency for ability to plug into a uStreamer.

Could we have all the methods of a `UTransport` be asynchronous, giving a bit more control over it to the implementor?

For the record, the `up-streamer-rust` implementation in PR uses the `up-rust` `UTransport` definition which has all async methods.